File: _designs_001/Read N Characters Given Read4 II/Solution.java
public class Solution {
    private static final int MAX_LEN = 512;

    public List<String> testReaderN(String s, StreamTokenizer tokenizer, List<Double> metrics) throws IOException {
        ReaderN readerN = new ReaderN(new Reader4(s));
        List<String> result = new ArrayList<String>();

        int cnt = Serializer.deserializeInt(tokenizer);
        for (int i = 0; i < cnt; i++) {
            char[] buffer = new char[MAX_LEN];
            int n = Serializer.deserializeInt(tokenizer);
            double startTime = System.nanoTime();
            int size = readerN.read(buffer, n);
            metrics.set(0, metrics.get(0) + (double) (System.nanoTime() - startTime) / 1000000L);
            result.add(new String(buffer, 0, size));
        }
        return result;
    }
}

File: _designs_001/Read N Characters Given Read4 II/Solution.rb
class Solution
  def testReaderN(s, inp, metrics)
    metrics[0] = 0
    reader_n, r = ReaderN.new(Reader4.new(s)), []
    Serializer.deserializeInt(inp).times do
      n, buf = Serializer.deserializeInt(inp), ''
      start_time = Time.now
      reader_n.read(buf, n)
      r << buf
      metrics[0] += (Time.now - start_time) * 1000
    end
    r
  end
end
